### ISSUE TITLE
LS Integration: Document syncing on diagnostics

### ISSUE THREAD

**Issue by kathrynkodama** [OpenLiberty/liberty-tools-intellij#85]
Document seems to fall out of sync, outdated diagnostics are still being shown after they are fixed.

For example, valid diagnostic:

<img width="902" alt="image" src="https://user-images.githubusercontent.com/26146482/187963545-6d7dfb47-8741-45f5-ae32-2329da2ed0a5.png">

After adding the corresponding method, the outdated diagnostic is still being shown:
<img width="985" alt="Screen Shot 2022-09-01 at 12 19 00 PM" src="https://user-images.githubusercontent.com/26146482/187963724-da056fbe-ddce-400d-8692-66b2d2bbc68b.png">

**Comment by turkeylurkey** [comment]
This is a clarification of the previous comment. When you first open the editor there is no diagnostic:
<img width="680" alt="image" src="https://user-images.githubusercontent.com/24705144/199616327-4a9937c3-8ba4-4c6d-9e6a-5f15cc5ef8aa.png">
Then hit one character and the diagnostic appears as expected, there is no one keystroke difference:
<img width="745" alt="image" src="https://user-images.githubusercontent.com/24705144/199617569-57102fd4-f5df-45e2-aa53-bf648755567b.png">
However there is no hover information at this point. One keystroke (backspace) later the diagnostic is rendered differently (all red):
<img width="938" alt="image" src="https://user-images.githubusercontent.com/24705144/199617770-0e2ee8b6-4233-477e-9187-2e5479ebe651.png">
And the hover information is one keystroke behind.

The diagnostic is generated when the language server receives the message microprofile/java/diagnostics as specified in MicroProfileJavaDiagnosticsProvider implemented in MicroProfileLanguageClient:
```
@Override
  public CompletableFuture<List<PublishDiagnosticsParams>> getJavaDiagnostics(MicroProfileJavaDiagnosticsParams javaParams) {
    return runAsBackground("Computing Java diagnostics", () -> PropertiesManagerForJava.getInstance().diagnostics(javaParams, PsiUtilsLSImpl.getInstance(getProject())));
  }
```
PropertiesManagerForJava collects the class name from an extension point:
```
List<IJavaDiagnosticsParticipant> definitions = IJavaDiagnosticsParticipant.EP_NAME.extensions()
```
see 
```
<extensionPoint name="javaASTValidator.validator" beanClass="com.langserver.devtools.intellij.lsp4mp4ij.psi.core.java.validators.JavaASTValidatorExtensionPointBean"/>
```
and calls the IntelliJ framework to walk the syntax tree and check for errors:
```
	private static void collectDiagnosticsInFile(JavaDiagnosticsContext context, List<Diagnostic> diagnostics) {
		// Collect the list of JavaASTValidator which are adapted for the current AST
		// compilation unit to validate.
		Collection<JavaASTValidator> validators = JavaASTValidatorRegistry.getInstance().getValidators(context, diagnostics);
		if (!validators.isEmpty()) {
			// Visit the AST compilation unit and process each validator.
			PsiFile ast = context.getASTRoot();
			ast.accept(new MultiASTVisitor(validators));
```
That is how MicroProfileFaultToleranceASTValidator.visitMethod() calls validateMethod(node).

**Comment by evie-lau** [comment]
From debugging both the diagnostics collection/publishing and the documentChanged actions, it seems like the diagnostics may be running before the document change has fully processed, so it is getting diagnostics for the pre-change state.

This also lines up with when initially making the fallbackMethod invalid, there is only a red underline for syntax, but no diagnostic information on hover or red text.

**Comment by evie-lau** [comment]
Documenting here that same issue occurs in the Quarkus tools plugin which also uses LSP4MP.

**Comment by turkeylurkey** [comment]
See this comment above: "the diagnostic is rendered differently (all red)". If you wait a minute and click on another window and then click back on IntelliJ the red diagnostic is gone and everything is fine (assuming you fixed the error as in the example above). IntelliJ clears the "intentions" after a while.

**Comment by evie-lau** [comment]
In my experience, even immediately switching windows (without needing to switch back to IntelliJ) caused IntelliJ to update the diagnostics.

**Comment by evie-lau** [comment]
One difference I noticed that might be worth investigating...
In LSP4Jakarta, the server capabilities is using `setTextDocumentSync(TextDocumentSyncKind.Incremental)`
In LCLS, the server capabilities is using `setTextDocumentSync(TextDocumentSyncKind.Full)` - I haven't seen syncing issues with LCLS

**Comment by kathrynkodama** [comment]
Just noting that I am seeing syncing issues with all of the language servers currently on main - LSP4Jakarta, LSP4MP, LemMinX + Liberty LemMinX ext, and Liberty Config LS

**Comment by dshimo** [comment]
Documenting here that LCLS diagnostics do not appear until interacted with, such as saving or typing

**Comment by kathrynkodama** [comment]
Noting that with #254, I see the following exception multiple times in the logs on startup:
```
com.intellij.openapi.progress.ProcessCanceledException: java.lang.Throwable
	at com.intellij.openapi.progress.util.AbstractProgressIndicatorBase.throwIfCanceled(AbstractProgressIndicatorBase.java:157)
	at com.intellij.openapi.progress.util.AbstractProgressIndicatorBase.checkCanceled(AbstractProgressIndicatorBase.java:147)
	at com.intellij.codeInsight.daemon.impl.DaemonProgressIndicator.checkCanceled(DaemonProgressIndicator.java:97)
	at com.intellij.openapi.progress.util.ProgressWrapper.checkCanceled(ProgressWrapper.java:82)
	at com.intellij.openapi.progress.impl.CoreProgressManager.doCheckCanceled(CoreProgressManager.java:143)
	at com.intellij.openapi.progress.ProgressManager.checkCanceled(ProgressManager.java:219)
	at com.intellij.serviceContainer.ContainerUtilKt.checkCanceledIfNotInClassInit(containerUtil.kt:16)
	at com.intellij.serviceContainer.ComponentManagerImpl.instantiateClass(ComponentManagerImpl.kt:839)
	at io.openliberty.tools.intellij.lsp4mp.lsp4ij.LanguageServersRegistry$ExtensionLanguageServerDefinition.createLanguageClient(LanguageServersRegistry.java:89)
	at io.openliberty.tools.intellij.lsp4mp.lsp4ij.LanguageServerWrapper.start(LanguageServerWrapper.java:215)
	at io.openliberty.tools.intellij.lsp4mp.lsp4ij.LanguageServiceAccessor.getLSWrapperForConnection(LanguageServiceAccessor.java:443)
	at io.openliberty.tools.intellij.lsp4mp.lsp4ij.LanguageServiceAccessor.getLSWrappers(LanguageServiceAccessor.java:321)
	at io.openliberty.tools.intellij.lsp4mp.lsp4ij.operations.diagnostics.LSPLocalInspectionTool.checkFile(LSPLocalInspectionTool.java:70)
	at com.intellij.codeInspection.LocalInspectionTool$1.visitFile(LocalInspectionTool.java:141)
	at com.intellij.extapi.psi.PsiFileBase.accept(PsiFileBase.java:60)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$10(InspectionRunner.java:317)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1154)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$11(InspectionRunner.java:325)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:174)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:215)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$12(InspectionRunner.java:331)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:159)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:151)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$13(InspectionRunner.java:294)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:130)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:119)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$14(InspectionRunner.java:294)
	at com.intellij.concurrency.JobLauncherImpl$1MyProcessQueueTask.lambda$call$0(JobLauncherImpl.java:297)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:608)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:683)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:639)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:607)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.concurrency.JobLauncherImpl$1MyProcessQueueTask.call(JobLauncherImpl.java:282)
	at com.intellij.concurrency.JobLauncherImpl$1MyProcessQueueTask.call(JobLauncherImpl.java:270)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1428)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.Throwable
	at com.intellij.openapi.util.ObjectTree.runWithTrace(ObjectTree.java:98)
	at com.intellij.openapi.util.ObjectTree.executeAll(ObjectTree.java:158)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:217)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:205)
	at com.intellij.codeInsight.daemon.impl.DaemonProgressIndicator.cancel(DaemonProgressIndicator.java:65)
Caused by: java.lang.Throwable

	at com.intellij.codeInsight.daemon.impl.DaemonCodeAnalyzerImpl.cancelUpdateProgress(DaemonCodeAnalyzerImpl.java:673)
	at com.intellij.codeInsight.daemon.impl.DaemonCodeAnalyzerImpl.stopProcess(DaemonCodeAnalyzerImpl.java:645)
	at com.intellij.codeInsight.daemon.impl.DaemonCodeAnalyzerImpl.stopProcessAndRestartAllFiles(DaemonCodeAnalyzerImpl.java:585)
	at com.intellij.codeInsight.daemon.impl.DaemonCodeAnalyzerImpl.restart(DaemonCodeAnalyzerImpl.java:579)
	at com.intellij.spellchecker.grazie.async.AsyncUtils.restartInspection(AsyncUtils.kt:21)
	at com.intellij.spellchecker.grazie.async.AsyncUtils$run$toRun$1$1$1.run(AsyncUtils.kt:38)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:779)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:730)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:724)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:749)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:918)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:766)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:450)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:791)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:449)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:105)
	at com.intellij.ide.IdeEventQueue.performActivity(IdeEventQueue.java:624)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:447)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:881)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:493)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
```

**Comment by kathrynkodama** [comment]
After the #267 fix, there is still some keystroke delay behaviour when switching from one file to another.
To reproduce:
1. Start `./gradlew runIde`
2. Open a Java file with some LSP4Jakarta problems. Slight delay waiting for the language servers to startup but without any window switching, keystrokes, problems view is correctly populated and we can see the inspections appear.
<img width="1559" alt="image" src="https://user-images.githubusercontent.com/26146482/219717599-cea9a054-0f31-46af-957a-fa82d6eb9d98.png">
3. Switch to a `server.xml` file with some expected diagnostics. Nothing appears in the Problems view and no underlining/marking on the UI. 
<img width="1221" alt="image" src="https://user-images.githubusercontent.com/26146482/219717697-33c8378e-13aa-4f4e-8d44-908dc866bb43.png">
4. Type one keystroke, diagnostics are populated as expected.
<img width="1236" alt="image" src="https://user-images.githubusercontent.com/26146482/219717756-2d3e7dfd-09ca-4479-b95a-58f08a3566ce.png">
I can reproduce with multiple LS, for example if I open a bootstrap.properties file instead of a server.xml, there is still the keystroke delay.

**Comment by kathrynkodama** [comment]
With the latest on main I can no longer reproduce the above comment's syncing error cc: @evie-lau  @dshimo

**Comment by evie-lau** [comment]
I'm still seeing some sync issues, but they are nowhere near as bad as before.
![image](https://user-images.githubusercontent.com/689163/221898536-deb97bf8-e8af-4ef1-bff1-c4234e046037.png)
Above, I deleted the `0`, allowed the diagnostic to show up (working as intended and on-time), then readded the `0`. The red underline disappeared, but the previously red-underlined text turns to red text, and the Problems view and hover both still show the outdated diagnostic. This seems to disappear when switching out of the IntelliJ window and back, or switching file tabs and back in the same subwindow (but not a side-by-side file in a different subwindow).

**Comment by evie-lau** [comment]
I added some logging to see the order of what was happening when it worked vs when it didn't (using old code before #277 fix to force `checkFile` when running `createMarkers`):
![image](https://user-images.githubusercontent.com/689163/221995449-ff0e3878-f5a1-4c09-8c2c-8719eba53469.png)

Seems like when it fell out of sync, `checkFile` had run before the updated diagnostics were accepted, so it kept the old problem descriptors.

**Comment by kathrynkodama** [comment]
Closing with final fix of #277 and opened https://github.com/OpenLiberty/liberty-tools-intellij/issues/278 as a future performance enhancement
